### ISSUE TITLE
Problem on waiting for tasks

### ISSUE THREAD

**Issue by rizwan-0** [WTTJ/algoliax#49]
### Environment

- Elixir version (elixir -v): Elixir 1.11.3 (compiled with Erlang/OTP 23)
- Algoliax version (mix deps): 0.6.0

### Background
I have created a helper function to wait for tasks to complete or 'published'. As the tasks are run asynchronously, I use the 'taskID' returned by the operation and check the task status on Algolia using Tesla. I am using this helper function in the test module, to test the filtering conditions on Algolia.

[Documentation](https://www.algolia.com/doc/rest-api/search/#get-a-tasks-status) for task status

My helper function:
```
def wait_algolia_task({:ok, %{"taskID" => task_id}} = algolia_response, index_name)
      when is_bitstring(index_name) do
    application_id = Application.fetch_env!(:algoliax, :application_id)
    api_key = Application.fetch_env!(:algoliax, :api_key)

    [
      {Tesla.Middleware.BaseUrl,
       "https://#{application_id}-dsn.algolia.net/1/indexes/#{index_name}"},
      {Tesla.Middleware.Headers, [{"X-Algolia-Application-Id", "#{application_id}"}]},
      {Tesla.Middleware.Headers, [{"X-Algolia-API-Key", "#{api_key}"}]},
      Tesla.Middleware.JSON
    ]
    |> Tesla.client()
    |> Tesla.get("/task/#{task_id}")
    |> parse_response(algolia_response, index_name)
  end

  defp parse_response(
         {:ok, %Tesla.Env{body: %{"status" => "published"}}},
         _algolia_resp,
         _index_name
       ),
       do: :ok

  defp parse_response(
         {:ok, %Tesla.Env{status: 200, body: %{"status" => "notPublished"}}},
         algolia_response,
         index_name
       ) do
    :timer.sleep(1000)
    wait_algolia_task(algolia_response, index_name)
  end

  defp parse_response(tesla_response, _algolia_response, _index_name),
    do: {:error, tesla_response}
```

An example of usage would be:
```
:ok = SomeModule.save_object(%{}) |> HelpersTest.wait_algolia_task("some_module")
```

### Expected behavior
I want is to move on to the next line only when the operations are finished _publishing_

### Actual behavior
After I wait for the operations, the next thing is to search the index with some specific filters and then assert the returned values. The asserting fails, which means that there were no search results, however, if I go to my Algolia dashboard and check for those values that I just uploaded, they are present on Algolia with the filters that I applied.
This happens not always but many times, sometimes on my local environment, and sometimes on the BitBucket pipeline.

**Comment by StephaneRob** [comment]
@rizwan-0 thanks for reaching out.
I'm not sure if this issue is related to Algoliax library. 
Do you have any other information? like, how do you make the assertion?

**Comment by rizwan-0** [comment]
@StephaneRob , thanks for the reply.
Maybe yes, I find it hard to believe that it is a problem with the Algoliax library. However, I needed some expert opinion from people using this library, so I thought this would be a good place to ask (maybe I'm wrong here).
Anyways, regarding the test, here is some more information on my issue.

My data_case file contains:
```
data_case.ex
  setup tags do
      :ok = Ecto.Adapters.SQL.Sandbox.checkout(RestWorld.Repo)
  
      unless tags[:async] do
        Ecto.Adapters.SQL.Sandbox.mode(RestWorld.Repo, {:shared, self()})
      end
  
      :ok
  end
```

Also in each test module, I'm using the setup_all also as:
```
one_of_the_test_module.ex
  setup_all do
    :ok = Ecto.Adapters.SQL.Sandbox.checkout(RestWorld.Repo)
    Ecto.Adapters.SQL.Sandbox.mode(RestWorld.Repo, :auto)
    
    # some entities defined here which would be used by every test in this module
  end
  
  test "testing update of an entity" do
    # create an entity
    {:ok, %{id: entity_id}} = create(%{})

    # update it
    {:ok, _updated_entity} = update(%{"id" => entity_id, "country" => "italy"})

    # upload the data to algolia
    :ok = 
      upload_to_algolia(entity)
      # returns taskID which is piped to helper_function to wait
      |> wait_algolia_task("entity_index")

    # run the index_entity() which queries algolia with the facetFilters
    {:ok, results} = index_entity(%{"country" => "italy"})
    # index_entity() is querying algolia with facetFilters. The returned result_ids are queryed on the DB and the final results are returned from the function

    assert length(results) >= 1

    result_ids = Enum.map(results, fn %{id: id} -> id end)
    # assert the entity_id in results
    assert entity_id in result_ids
  end
end
```

The index_entity() works fine as I tested it in an iex shell, but with these tests, sometimes it fails with no results found on algolia and sometimes it passes the test.

Quick question:
Is it possible that the endpoint on Algolia to check the status of a task is somehow the problem? Like giving me the "published" status when its not finished/published?

**Comment by StephaneRob** [comment]
Normally the `published` status means that the job is done on algolia side.
Did you try to add a little before running the search? 

btw; we will definitely add the wait task feature on the library 👍

**Comment by StephaneRob** [comment]
Can you try to add 500ms (or less) delay before executing the search?

**Comment by rizwan-0** [comment]
I did add a sleep of 500ms but the problem persists. I'm starting to think that the problem is with my testing module.
Anyways, I will keep exploring this issue and will update this thread if I find anything worth sharing.
Thank you for the suggestions @StephaneRob ..:thumbsup:

**Comment by StephaneRob** [comment]
@rizwan-0 wait task is coming in PR #51.

**Comment by StephaneRob** [comment]
Closed by #51
